### PR TITLE
Fix evaluation of cells referring to themselves

### LINKED ISSUE #1347: How to handle StackOverflowException in ClosedXML
## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug

**Version of ClosedXML**
0.94.2.0

**What is the current behavior?**
I have written code to check cell formula of excel cell in C# ClosedXML. Everything works fine, but only problem is, if the checked cell formula contains self cell then it gives StackOverflowException and this exception stops my service.

**What is the expected behavior or new feature?**
Please Handle StackOverflowException for self formulated cell.

**Did this work in previous versions of our tool?  Which versions?**
No

## Reproducibility

**Code to reproduce problem:**
```c#
var _xlWorkBook = new ClosedXML.Excel.XLWorkbook(fileStream);
var workSheet = xlWorkBook.Worksheet(1);
var row = workSheet.Row(2);
string cellValue = row.Cell(2).Value.ToString();
```
-
[Test File.xlsx](https://github.com/ClosedXML/ClosedXML/files/4179529/Test.File.xlsx)
 [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)

### PR DESCRIPTION
**Pull request by Pankraty**
Fixes #1347

### CODE REVIEW

File: ClosedXML_Tests/Excel/Cells/XLCellTests.cs [hunk 1]
```diff
@@ -974,6 +974,7 @@ public void InvalidFormulaShiftProducesREF()
             }
         }
 
+        [Test]
```
Comment (igitur) [inline ClosedXML_Tests/Excel/Cells/XLCellTests.cs]: 😩

Comment (igitur) [reply ClosedXML_Tests/Excel/Cells/XLCellTests.cs]: This was my fault.

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 2]
```diff
@@ -441,16 +441,22 @@ private object RecalculateFormula(string fA1)
...
-            {
-                var referenceCell = Worksheet.Workbook.Worksheet(sName).Cell(cAddress);
-                if (referenceCell.IsEmpty(XLCellsUsedOptions.AllContents))
-                    return 0;
-                else
-                    return referenceCell.Value;
+            if (Worksheet.Workbook.Worksheets.Contains(sName)
+                && XLHelper.IsValidA1Address(cAddress))
+            {
+                try
```
Comment (igitur) [inline ClosedXML/Excel/Cells/XLCell.cs]: I know that `try/catch` blocks severely affect performance. But does `try/finally` without `catch` perform well?

Comment (Pankraty) [reply ClosedXML/Excel/Cells/XLCell.cs]: From what I know, the most performance hit is caused by unwinding stack trace so without the exception there should be no difference. But you've raised an interesting question, i'll try to do a little of benchmarking.

Comment (Pankraty) [reply ClosedXML/Excel/Cells/XLCell.cs]: https://stackoverflow.com/a/4107320

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 3]
```diff
@@ -441,16 +441,22 @@ private object RecalculateFormula(string fA1)
...
-                w => String.Compare(w.Name, sName, true) == 0)
-                && XLHelper.IsValidA1Address(cAddress)
-                )
-            {
-                var referenceCell = Worksheet.Workbook.Worksheet(sName).Cell(cAddress);
-                if (referenceCell.IsEmpty(XLCellsUsedOptions.AllContents))
-                    return 0;
-                else
-                    return referenceCell.Value;
+            if (Worksheet.Workbook.Worksheets.Contains(sName)
```
Comment (igitur) [inline ClosedXML/Excel/Cells/XLCell.cs]: Please make this change (to use `.Contains` at line 470 too). Bonus points if you can put this change in a separate commit.

Comment (igitur) [reply ClosedXML/Excel/Cells/XLCell.cs]: Hmm, I see wee have many cases where we use `String.Compare(String, String, Boolean)` (not only for worksheet names, but also for named ranges). Maybe this warrants a new PR?

Comment (Pankraty) [reply ClosedXML/Excel/Cells/XLCell.cs]: I've extracted these changes in a separate commit and replaced all `.Any(x => String.Compare(...))` with the direct access, where applicable.

**Comment by igitur** [comment]
Is it ok if we base this on #1357  ?

**Comment by Pankraty** [comment]
> Is it ok if we base this on #1357 ?

Sure!

**Comment by igitur** [comment]
#1357 has been merged.  Please rebase this on `develop`.

**Comment by Pankraty** [comment]
Done

**Pull request merged**